lib/auth/auth_with_roles.go

```diff
@@ -6423,9 +6451,9 @@ func (a *ServerWithRoles) UpsertWindowsDesktop(ctx context.Context, s types.Wind
 		return trace.Wrap(err)
 	}
 
-	if s.GetHostID() == "" {
-		// dont try to insert desktops with empty hostIDs
-		return nil
+	// Ensure the host ID is valid.
+	if err := a.windowsDesktopHostIDExistsAndIsValid(ctx, s); err != nil {
```

Comment:
Every upsert now has to make a network call to validate a host ID. If we're doing LDAP discovery and on every discovery loop we upsert a couple thousand desktops, that's a couple thousand extra lookups against the backend.

This is particularly wasteful because all of these upserts would likely be with the same host ID, so we're repeating the same check many times in a short period.

Should we use `FnCache` to memoize the results of this call for a few minutes?

Reply:
IIUC the auth server uses an in-memory cache (unless explicitly disabled), thus even with thousands of desktops this won't cause performance issues.

Reply:
Do we know that or are we guessing? When it comes to performance, our instincts are often wrong. Have we done any tests or benchmarks or are we just assuming that the cache will save us?

Reply:
I've added a `tctl bench desktop` in TODO

Run against this branch, I get
```
$ tctl bench desktop --desktops 1000 --runs 3
# ...
Average creation time for 1000 desktops over 3 runs: 930.255972ms

$ tctl bench desktop --desktops 10000 --runs 3
# ...
Average creation time for 10000 desktops over 3 runs: 44.054836375s
```

Run against `master`, I get

```
$ tctl bench desktop --desktops 1000 --runs 3
# ...
Average creation time for 1000 desktops over 3 runs: 889.181736ms

$ tctl bench desktop --desktops 10000 --runs 3
# ...
Average creation time for 10000 desktops over 3 runs: 42.30581625s
```

which means a 4-5% increase in time per `UpsertWindowsDesktop`, which I'd consider negligible for this type of task.